<?php

namespace OpenExam\Models;

use Exception;
use OpenExam\Tests\Phalcon\TestModel;

/**
 * @author Anders Lövgren (Computing Department at BMC, Uppsala University)
 */
class ResourceModel extends Resource
{

        public function initialize()
        {
                parent::initialize();
        }

}

/**
 * Generated by PHPUnit_SkeletonGenerator 1.2.1 on 2014-09-21 at 15:11:59.
 * @author Anders Lövgren (Computing Department at BMC, Uppsala University)
 */
class ResourceTest extends TestModel
{

        public function __construct()
        {
                parent::__construct(new ResourceModel());
        }

        /**
         * @group model
         */
        public function testRelations()
        {
                $resource = Resource::findFirst();
                self::assertNotNull($resource);

                self::assertNotEquals($resource->exam->count(), 0);
                self::assertTrue(count($resource->exam) == 1);
        }

        /**
         * @group model
         */
        public function testProperties()
        {
                $values = array(
                        'exam_id' => Exam::findFirst()->id,
                        'name'    => 'Name1',
                        'path'    => '/tmp/path',
                        'type'    => 'video',
                        'subtype' => 'mp4',
                        'user'    => 'user1'
                );

                try {
                        $helper = new TestModel(new Resource());
                        $helper->tryPersist();
                        self::fail("Excepted constraint violation exception");
                } catch (Exception $exception) {
                        // Expected exception
                }

                try {
                        $helper = new TestModel(new Resource());
                        $helper->tryPersist($values);
                } catch (Exception $exception) {
                        self::fail($exception);
                }

                $helper->checkDefaults(array(
                        'shared' => Resource::SHARED_EXAM
                ));

                $values = array(
                        'exam_id'      => Exam::findFirst()->id,
                        'name'         => 'Name1',
                        'path'         => '/tmp/path',
                        'type'         => 'video',
                        'subtype'      => 'mp4',
                        'user'         => 'user1',
                        'non_existing' => 666   // ignored wihout error
                );
                try {
                        $helper = new TestModel(new Resource());
                        $helper->tryPersist($values);
                } catch (Exception $exception) {
                        self::fail("Unexcepted constraint violation exception");
                }
        }

        /**
         * @group model
         */
        public function testDefaults()
        {
                $values = array(
                        'exam_id' => Exam::findFirst()->id,
                        'name'    => 'Name1',
                        'path'    => '/tmp/path',
                        'type'    => 'video',
                        'subtype' => 'mp4',
                        'user'    => 'user1',
                );
        }

        /**
         * @covers OpenExam\Models\Contributor::getSource
         * @group model
         */
        public function testGetSource()
        {
                $expect = "resources";
                $actual = $this->object->getSource();
                self::assertNotNull($actual);
                self::assertEquals($expect, $actual);
        }

}
